be/src/vec/exec/distinct_vaggregation_node.cpp

```diff
@@ -0,0 +1,159 @@
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "vec/exec/distinct_vaggregation_node.h"
+
+#include "runtime/runtime_state.h"
+#include "vec/aggregate_functions/aggregate_function_uniq.h"
+#include "vec/exec/vaggregation_node.h"
+
+namespace doris {
+class ObjectPool;
+} // namespace doris
+
+namespace doris::vectorized {
+
+DistinctAggregationNode::DistinctAggregationNode(ObjectPool* pool, const TPlanNode& tnode,
+                                                 const DescriptorTbl& descs)
+        : AggregationNode(pool, tnode, descs) {}
+
+Status DistinctAggregationNode::get_next(RuntimeState* state, Block* block, bool* eos) {
+    SCOPED_TIMER(_runtime_profile->total_time_counter());
+
+    if (_is_streaming_preagg) {
+        RETURN_IF_CANCELLED(state);
+        release_block_memory(_preagg_block);
+        while (_preagg_block.rows() == 0 && !_child_eos) {
+            RETURN_IF_ERROR(_children[0]->get_next_after_projects(
+                    state, &_preagg_block, &_child_eos,
+                    std::bind((Status(ExecNode::*)(RuntimeState*, vectorized::Block*, bool*)) &
+                                      ExecNode::get_next,
+                              _children[0], std::placeholders::_1, std::placeholders::_2,
+                              std::placeholders::_3)));
+        };
+        if (_preagg_block.rows() != 0) {
+            RETURN_IF_ERROR(do_pre_agg(&_preagg_block, block));
+        }
+        if (_preagg_block.rows() == 0 || reached_limit()) {
+            *eos = true;
+        }
+    } else {
+        RETURN_IF_ERROR(pull(state, block, eos));
+    }
+    return Status::OK();
+}
+
+Status DistinctAggregationNode::_distinct_pre_agg_with_serialized_key(
+        doris::vectorized::Block* in_block, doris::vectorized::Block* out_block) {
+    SCOPED_TIMER(_build_timer);
+    DCHECK(!_probe_expr_ctxs.empty());
+
+    size_t key_size = _probe_expr_ctxs.size();
+    ColumnRawPtrs key_columns(key_size);
+    {
+        SCOPED_TIMER(_expr_timer);
+        for (size_t i = 0; i < key_size; ++i) {
+            int result_column_id = -1;
+            RETURN_IF_ERROR(_probe_expr_ctxs[i]->execute(in_block, &result_column_id));
+            in_block->get_by_position(result_column_id).column =
+                    in_block->get_by_position(result_column_id)
+                            .column->convert_to_full_column_if_const();
+            key_columns[i] = in_block->get_by_position(result_column_id).column.get();
+        }
+    }
+
+    int rows = in_block->rows();
+    IColumn::Selector distinct_row;
+    distinct_row.reserve(rows);
+
+    RETURN_IF_CATCH_EXCEPTION(
+            _emplace_into_hash_table_to_distinct(distinct_row, key_columns, rows));
+
+    bool mem_reuse = _make_nullable_keys.empty() && out_block->mem_reuse();
+    if (mem_reuse) {
```

Comment:
should use build_mutable_block to replace the code

Reply:
have some problem to use the function,  maybe keep this as agg node  [                      if (!mem_reuse) {](https://github.com/apache/doris/blob/527547b4ed14cdf524576755961952119f12bc06/be/src/vec/exec/vaggregation_node.cpp#L1121)